File: src/components/timezone-search.ts
import { searchTimezones } from '../services/timezone-search';
import { getClocks, addClock } from '../state/clocks';
import { subscribe } from '../state/event-bus';
import type { TimezoneMetadata, TimeZoneEntry } from '../types/timezone';

export class TimezoneSearchModal {
  private element: HTMLElement;
  private input: HTMLInputElement;
  private resultsContainer: HTMLElement;
  private warningElement: HTMLElement;
  
  private results: TimezoneMetadata[] = [];
  private selectedIndex: number = -1;

  constructor() {
    this.element = document.createElement('div');
    this.element.className = 'search-modal';
    this.element.style.display = 'none';

    const content = document.createElement('div');
    content.className = 'search-modal__content';

    this.input = document.createElement('input');
    this.input.type = 'text';
    this.input.className = 'search-modal__input';
    this.input.placeholder = 'Search timezone (e.g., Tokyo, UTC+9)';
    
    this.warningElement = document.createElement('div');
    this.warningElement.className = 'search-modal__warning';
    this.warningElement.style.display = 'none';

    this.resultsContainer = document.createElement('ul');
    this.resultsContainer.className = 'search-modal__results';

    content.appendChild(this.input);
    content.appendChild(this.warningElement);
    content.appendChild(this.resultsContainer);
    this.element.appendChild(content);

    document.body.appendChild(this.element);

    this.bindEvents();
    
    subscribe('OPEN_SEARCH_MODAL', () => {
      this.open();
    });
  }

  private bindEvents() {
    document.addEventListener('keydown', (e) => {
      if ((e.ctrlKey || e.metaKey) && (e.key === 'k' || e.key === 'n')) {
        e.preventDefault();
        this.open();
      }
    });

    this.element.addEventListener('click', (e) => {
      if (e.target === this.element) {
        this.close();
      }
    });

    this.input.addEventListener('input', () => {
      this.handleSearch();
    });

    this.input.addEventListener('keydown', (e) => {
      if (e.key === 'Escape') {
        this.close();
      } else if (e.key === 'ArrowDown') {
        e.preventDefault();
        this.moveSelection(1);
      } else if (e.key === 'ArrowUp') {
        e.preventDefault();
        this.moveSelection(-1);
      } else if (e.key === 'Enter') {
        e.preventDefault();
        this.addSelected();
      }
    });
  }

  private open() {
    this.element.style.display = 'flex';
    this.input.value = '';
    this.results = [];
    this.selectedIndex = -1;
    this.warningElement.style.display = 'none';
    this.renderResults();
    this.input.focus();
  }

  private close() {
    this.element.style.display = 'none';
  }

  private handleSearch() {
    const query = this.input.value;
    this.results = searchTimezones(query);
    this.selectedIndex = this.results.length > 0 ? 0 : -1;
    this.warningElement.style.display = 'none';
    this.renderResults();
  }

  private moveSelection(delta: number) {
    if (this.results.length === 0) return;
    this.selectedIndex += delta;
    if (this.selectedIndex < 0) this.selectedIndex = 0;
    if (this.selectedIndex >= this.results.length) this.selectedIndex = this.results.length - 1;
    this.renderResults();
  }

  private renderResults() {
    this.resultsContainer.innerHTML = '';
    this.results.forEach((tz, index) => {
      const li = document.createElement('li');
      li.className = 'search-modal__result-item';
      if (index === this.selectedIndex) {
        li.classList.add('selected');
      }
      
      const offsetHours = tz.utc_offset_minutes / 60;
      const offsetStr = offsetHours >= 0 ? `+${offsetHours}` : `${offsetHours}`;
      
      li.textContent = `${tz.city || tz.id} (${tz.country || 'N/A'}) UTC${offsetStr}`;
      li.addEventListener('click', () => {
        this.selectedIndex = index;
        this.addSelected();
      });
      li.addEventListener('mouseenter', () => {
        this.selectedIndex = index;
        this.renderResults();
      });
      this.resultsContainer.appendChild(li);
    });
  }

  private async addSelected() {
    if (this.selectedIndex >= 0 && this.selectedIndex < this.results.length) {
      const selectedTz = this.results[this.selectedIndex];
      const existingClocks = getClocks();
      
      if (existingClocks.some(c => c.iana_id === selectedTz.id)) {
        this.warningElement.textContent = 'This timezone is already in your list.';
        this.warningElement.style.display = 'block';
        return;
      }
      
      if (existingClocks.length >= 30) {
        this.warningElement.textContent = 'Maximum of 30 timezones reached.';
        this.warningElement.style.display = 'block';
        return;
      }

      const newClock: TimeZoneEntry = {
        id: crypto.randomUUID(),
        iana_id: selectedTz.id,
        display_label: selectedTz.city || selectedTz.id,
        sort_order: existingClocks.length,
        pinned_to_tray: false,
        created_at: Date.now()
      };

      await addClock(newClock);
      this.close();
    }
  }
}

File: src-ui/src/components/timezone_search.rs
use leptos::*;
use crate::state::{ClockState, ModalState};
use crate::models::{TimeZoneEntry, TimezoneMetadata};
use crate::utils::TIMEZONES;
use chrono::Utc;

#[derive(Clone, PartialEq)]
struct SearchResult {
    tz: TimezoneMetadata,
    match_name: String,
}

#[component]
pub fn TimezoneSearchModal() -> impl IntoView {
    let modal_state = use_context::<ModalState>().expect("ModalState missing");
    let clock_state = use_context::<ClockState>().expect("ClockState missing");

    let (search_query, set_search_query) = create_signal("".to_string());
    let (selected_index, set_selected_index) = create_signal(0_usize);
    let (warning, set_warning) = create_signal("".to_string());

    let results = create_memo(move |_| {
        let query = search_query.get().to_lowercase();
        if query.is_empty() {
            return vec![];
        }

        let mut matched = Vec::new();
        TIMEZONES.with(|map| {
            for (_, tz) in map.iter() {
                let mut match_name = None;
                
                if let Some(city) = &tz.city {
                    if city.to_lowercase().contains(&query) {
                        match_name = Some(city.clone());
                    }
                }
                
                if match_name.is_none() {
                    for alias in &tz.aliases {
                        if alias.to_lowercase().contains(&query) {
                            match_name = Some(alias.clone());
                            break;
                        }
                    }
                }
                
                if match_name.is_none() && tz.id.to_lowercase().contains(&query) {
                    match_name = Some(tz.id.clone());
                }
                
                if match_name.is_none() {
                    if let Some(country) = &tz.country {
                        if country.to_lowercase().contains(&query) {
                            match_name = tz.city.clone().or_else(|| Some(tz.id.clone()));
                        }
                    }
                }
                
                if let Some(name) = match_name {
                    matched.push(SearchResult {
                        tz: tz.clone(),
                        match_name: name,
                    });
                }
            }
        });

        // Sort results to be deterministic
        matched.sort_by(|a, b| a.match_name.cmp(&b.match_name));
        matched.truncate(50); // limit
        
        if matched.is_empty() {
            set_selected_index.set(0);
        } else {
            // clamp
            let idx = selected_index.get_untracked();
            if idx >= matched.len() {
                set_selected_index.set(matched.len() - 1);
            }
        }
        
        matched
    });

    let is_open = move || modal_state.show_timezone_search.get();
    let search_state = move || crate::product::timezone_search_copy(
        &search_query.get(),
        results.get().len(),
    );

    let close_modal = move || {
        modal_state.show_timezone_search.set(false);
        set_search_query.set("".into());
        set_warning.set("".into());
    };

    let add_selected = move |idx: usize| {
        let res = results.get_untracked();
        if idx < res.len() {
            let result = &res[idx];
            let mut clocks = clock_state.zones.get_untracked();
            
            if clocks.iter().any(|c| c.iana_id == result.tz.id) {
                set_warning.set(format!("{} is already in your clock list.", result.tz.id));
                return;
            }

            if clocks.len() >= 30 {
                set_warning.set("Maximum of 30 timezones reached.".into());
                return;
            }

            clocks.push(TimeZoneEntry {
                id: uuid::Uuid::new_v4().to_string(),
                iana_id: result.tz.id.clone(),
                display_label: result.match_name.clone(),
                sort_order: clocks.len(),
                pinned_to_tray: false,
                created_at: Utc::now().timestamp_millis() as u64,
            });

            crate::store::apply_zones(clock_state, clocks.clone());
            spawn_local(async move {
                crate::store::save_zones(clocks).await;
            });
            close_modal();
        }
    };

    view! {
        <div class="search-modal" style=move || if is_open() { "display: flex;" } else { "display: none;" }>
            <div class="search-modal__content">
                <input 
                    type="text" 
                    class="search-modal__input" 
                    placeholder="Search timezone (e.g., Tokyo, Houston, UTC+9)..."
                    prop:value=search_query
                    on:input=move |ev| {
                        set_search_query.set(event_target_value(&ev));
                        set_warning.set("".into());
                        set_selected_index.set(0);
                    }
                    on:keydown=move |ev| {
                        let key = ev.key();
                        if key == "Escape" {
                            close_modal();
                        } else if key == "ArrowDown" {
                            ev.prevent_default();
                            let len = results.get_untracked().len();
                            let mut idx = selected_index.get_untracked();
                            if len > 0 && idx < len - 1 {
                                idx += 1;
                                set_selected_index.set(idx);
                            }
                        } else if key == "ArrowUp" {
                            ev.prevent_default();
                            let mut idx = selected_index.get_untracked();
                            if idx > 0 {
                                idx -= 1;
                                set_selected_index.set(idx);
                            }
                        } else if key == "Enter" {
                            ev.prevent_default();
                            add_selected(selected_index.get_untracked());
                        }
                    }
                />
                
                <div class="search-modal__warning" style=move || if warning.get().is_empty() { "display: none;" } else { "display: block;" }>
                    {move || warning.get()}
                </div>

                <div style="margin-bottom: 0.85rem; padding: 0.85rem 0.95rem; border-radius: 0.9rem; background: rgba(148, 163, 184, 0.08); border: 1px solid rgba(148, 163, 184, 0.18);">
                    <div style="font-size: 0.96rem; font-weight: 700; margin-bottom: 0.25rem;">
                        {move || search_state().title}
                    </div>
                    <div style="color: var(--theme-muted); line-height: 1.45; font-size: var(--font-size-sm);">
                        {move || search_state().body}
                    </div>
                </div>

                <ul class="search-modal__results">
                    <For
                        each=move || results.get().into_iter().enumerate()
                        key=|(i, r)| format!("{}-{}", r.tz.id, i)
                        children=move |(idx, result)| {
                            let offset_hours = result.tz.utc_offset_minutes as f32 / 60.0;
                            let offset_str = if offset_hours >= 0.0 {
                                format!("+{}", offset_hours)
                            } else {
                                format!("{}", offset_hours)
                            };

                            let is_selected = move || selected_index.get() == idx;
                            let class_name = move || {
                                if is_selected() {
                                    "search-modal__result-item selected"
                                } else {
                                    "search-modal__result-item"
                                }
                            };

                            let match_name = result.match_name.clone();

                            view! {
                                <li 
                                    class=class_name
                                    on:click=move |_| {
                                        set_selected_index.set(idx);
                                        add_selected(idx);
                                    }
                                    on:mouseenter=move |_| {
                                        set_selected_index.set(idx);
                                    }
                                >
                                    <span style="font-weight: var(--font-weight-bold); margin-right: var(--spacing-sm);">
                                        {match_name}
                                    </span>
                                    <span style="color: var(--theme-muted); font-size: var(--font-size-sm);">
                                        {format!("({}) UTC{}", result.tz.country.unwrap_or_else(|| "N/A".into()), offset_str)}
                                    </span>
                                </li>
                            }
                        }
                    />
                </ul>
            </div>
            
            <div 
                style="position: absolute; top: 0; left: 0; right: 0; bottom: 0; z-index: -1;" 
                on:click=move |_| close_modal()
            ></div>
        </div>
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn timezone_search_copy_covers_idle_and_results_states() {
        let idle = crate::product::timezone_search_copy("", 0);
        assert!(idle.title.contains("Search by city"));

        let ready = crate::product::timezone_search_copy("paris", 2);
        assert_eq!(ready.title, "2 matches ready");
    }
}
